app/src/ingest_bem_pdfs.py

```diff
@@ -76,15 +79,62 @@ def _parse_pdf(file: BinaryIO) -> list[EnrichedText]:
     for text in grouped_texts:
         text.id = str(uuid.uuid1())
     assert len(set(text.id for text in grouped_texts)) == len(grouped_texts)
+
     return grouped_texts
 
 
 def enrich_texts(file: BinaryIO) -> list[EnrichedText]:
+    unstuctured_elem_list = partition_pdf(file=file, strategy="fast")
+    enrich_text_list = []
     "Placeholder function. Will be implemented for DST-414, probably in a different file."
     outline: list[Heading] = pdf_utils.extract_outline(file)
-    for _heading in outline:
-        pass
-    return []
+    current_headings = []
+    for element in unstuctured_elem_list:
+        if element.category == "Footer" or element.category == "Header":
+            continue
+
+        # Unstructured fails to categorize the date strings in the header,
+        # so manually check for that and ignore those too
+        if element.category == "UncategorizedText" and re.match(
+            r"^\d{1,2}-\d{1,2}-\d{4}$", element.text
+        ):
+            continue
+
+        if element.category == "Title":
+            current_headings = _get_current_heading(outline, element, current_headings)
+        enriched_text_item = EnrichedText(
+            text=element.text,
+            type=element.category,
+            page_number=element.metadata.page_number,
+            headings=current_headings,
+            id=element._element_id,
+        )
+
+        enrich_text_list.append(enriched_text_item)
+    return enrich_text_list
+
+
+def match_heading(
+    outline: list[Heading], heading_name: str, page_number: int | None
+) -> Heading | None:
+    for heading in outline:
+        if heading_name.lower() in heading.title.lower() and heading.pageno == page_number:
+            return heading
+    return None
+
+
+def _get_current_heading(outline, element, current_headings):
+    if heading := match_heading(outline, element.text, element.metadata.page_number):
+        if heading.level == 1:
+            current_headings = [heading]
+        else:
+            if heading.title != current_headings[-1].title:
+                if len(current_headings) == heading.level:
+                    current_headings = current_headings[:-1]
```

Comment:
I don't think this is handling the case where you move from a heading level 3 to a level 1. Line 133 is only removing/popping one level. See https://github.com/navapbc/labs-decision-support-tool/blob/886b654922b62a7ad64aa11a1b3296dfe528c0f7/app/src/ingestion/pdf_stylings.py#L157-L159

Reply:
this should only be checking the case where the heading levels are the same and replacing the current level. ex: replacing  Heading(title='Family Independence Program (FIP)', level=2, pageno=1) with Heading(title='Program Goal', level=2, pageno=1); I'll add a case to check for the other level jump